### ISSUE TITLE
Connection Disconnect while processing request

### ISSUE THREAD

**Issue by sandeepjangir** [esl/MongooseIM#2022]
MongooseIM version: 3.0
Installed from: source
Erlang/OTP version: comes with 3.0

Getting session disconnection with these logs:

```
15:32:15.620 [debug] sending to process <0.1613.0>
15:32:15.620 [debug] routing done
15:32:15.740 [info] Timeout on ejabberd_tls:send
15:32:15.740 [info] ({socket_state,ejabberd_tls,{ejabberd_tls_socket,fast_tls,#Port<0.23255>,{tlssock,#Port<0.23255>,#Ref<0.3499729927.131727362.16712>},[verify_none,{certfile,[112,114,105,118,47,115,115,108,47,110,103,97,103,101,97,112,112,46,112,101,109]},{protocol_options,[110,111,95,115,115,108,118,51]}],false},<0.1612.0>}) Close session for sandeep@mg.myapp.com/Sandeeps-MacBook-Pro
15:32:15.740 [debug] rerouting unacked messages
15:32:15.741 [debug] route
        from {jid,<<"sandeep">>,<<"mg.myapp.com">>,<<>>,<<"sandeep">>,<<"mg.myapp.com">>,<<>>}
        to {jid,<<"sandeep">>,<<"mg.myapp.com">>,<<"Sandeeps-MacBook-Pro">>,<<"sandeep">>,<<"mg.myapp.com">>,<<"Sandeeps-MacBook-Pro">>}
        packet #{attrs => [{<<"id">>,<<"purple1112ae65">>},{<<"type">>,<<"result">>}],element => {xmlel,<<"iq">>,[{<<"id">>,<<"purple1112ae65">>},{<<"type">>,<<"result">>}],[{xmlel,<<"query">>,[{<<"xmlns">>,<<"jabber:iq:roster">>}],[{xmlel,<<"item">>,[{<<"subscription">>,<<"both">>},{<<"jid">>,<<"prasanna@im.ngageapp.com">>}],[]},{xmlel,<<"item">>,[{<<"subscription">>,<<"both">>},{<<"jid">>,<<"918128184101@im.ngageapp.com">>}],[]},{xmlel,<<"item">>,[{<<"subscription">>,<<"both">>},{<<"jid">>,<<"9983@mg.myapp.com">>}],[]},{xmlel,<<"item">>,[{<<"subscription">>,<<"both">>},{<<"jid">>,<<"prasanna@mg.myapp.com">>}],[]},{xmlel,<<"item">>,[{<<"ask">>,<<"subscribe">>},{<<"subscription">>,<<"none">>},{<<"jid">>,<<"6-6001_113-otr@mg.myapp.com">>}],[]},{xmlel,<<"item">>,[{<<"subscription">>,<<"none">>},{<<"jid">>,<<"6-6001_72@im.ngageapp.com">>}],[]},{xmlel,<<"item">>,[{<<"ask">>,<<"subscribe">>},{<<"subscription">>,<<"none">>},{<<"jid">>,<<"6-6001_115@mg.myapp.com">>}],[]},{xmlel,<<"item">>,[{<<"subscription">>,<<"both">>},{<<"name">>,<<"Vijay Vala">>},{<<"jid">>,<<"918128184102@abc.myapp.com">>}],[{xmlel,<<"group">>,[],[{xmlcdata,<<"Buddies">>}]}]},{xmlel,<<"item">>,[{<<"ask">>,<<"subscribe">>},{<<"subscription">>,<<"from">>},{<<"jid">>,<<"918128184101@abc.myapp.com">>}],[{xmlel,<<"group">>,[],[{xmlcdata,<<"Buddies">>}]}]},{xmlel,<<"item">>,[{<<"ask">>,<<"subscribe">>},{<<"subscription">>,<<"from">>},{<<"jid">>,<<"918128184101@mg.myapp.com">>}],[]}]}]},from => <<"sandeep@mg.myapp.com/Sandeeps-MacBook-Pro">>,from_jid => {jid,<<"sandeep">>,<<"mg.myapp.com">>,<<"Sandeeps-MacBook-Pro">>,<<"sandeep">>,<<"mg.myapp.com">>,<<"Sandeeps-MacBook-Pro">>},mongoose_acc => true,name => <<"iq">>,persistent_properties => [{origin_jid,{jid,<<"sandeep">>,<<"mg.myapp.com">>,<<"Sandeeps-MacBook-Pro">>,<<"sandeep">>,<<"mg.myapp.com">>,<<"Sandeeps-MacBook-Pro">>}},{origin_sid,{{1533,655934,403275},<0.1613.0>}}],ref => #Ref<0.3499729927.131596289.25335>,timestamp => {1533,655935,613004},to_jid => {jid,<<"sandeep">>,<<"mg.myapp.com">>,<<>>,<<"sandeep">>,<<"mg.myapp.com">>,<<>>},type => <<"result">>}
```

May I know what the reason of disconnection?

Thanks
Sandeep

**Comment by sandy5487** [comment]
One more error log found

`2018-08-07 17:54:30.145 [error] <0.1776.0>@ejabberd_c2s:handle_incoming_message:1314 Unexpected info: {#Ref<0.102715387.3910926338.97284>,{ok,[{offline_msg,{<<"918128184101">>,<<"im.myapp.com">>},{1533,664397,553089},never,{jid,<<"6-6001_115">>,<<"im.myapp.com">>,<<"A-6001-357871051876640">>,<<"6-6001_115">>,<<"im.myapp.com">>,<<"A-6001-357871051876640">>},{jid,<<"918128184101">>,<<"im.myapp.com">>,<<>>,<<"918128184101">>,<<"im.myapp.com">>,<<>>},{xmlel,<<"message">>,[{<<"from">>,<<"6-6001_115@im.myapp.com/A-6001-357871051876640">>},{<<"to">>,<<"918128184101@im.myapp.com">>},{<<"xml:lang">>,<<"en">>},{<<"id">>,<<"ExIDg-897">>},{<<"chat_sent">>,<<"2018-08-07T17:45:52.515Z">>},{<<"type">>,<<"chat">>}],[{xmlel,<<"ca">>,[],[]},{xmlel,<<"app_id">>,[],[{xmlcdata,<<"6001">>}]},{xmlel,<<"yek">>,[],[{xmlcdata,<<"Dht9dfIOTUtjMBVxsejN3zt6gANRS4UBgq4A0MQeefII7C0ri1JvGogch9k2HMyxkAU1+L7rTLEQTPnQvXM6WEr9jtFrTRBZUiJFJOcDNWYcPnFk0H6Vw1RDHz5UEuI5tKZBY3BofnXIsJMf2GZYPJBnwgAgBnD3PqfbDVuaIiQ=">>}]},{xmlel,<<"yek_ym">>,[],[{xmlcdata,<<"1RAJiv67dLfTVjdqfk4+KGfyaTC0C1nh9P3HAvu45DV+Wv1+CZ3fFXWBPmOmbCSrW0tsqWwuWAvzlBlQxX0H+B6PBX7cEg9bIC1DbaQnFT7cLAzVlXmyVOW8RigsdNCLaxAa8rzU7Xa19NnShOwgZvfEGz99UzAvqQiDENm4XeM=">>}]},{xmlel,<<"subject">>,[],[{xmlcdata,<<"poke">>}]},{xmlel,<<"body">>,[],[{xmlcdata,<<"+VXCAZy7yl9lYyRAWYQpaA==">>}]},{xmlel,<<"request">>,[{<<"xmlns">>,<<"urn:xmpp:receipts">>}],[]},{xmlel,<<"stanza-id">>,[{<<"by">>,<<"918128184101@im.myapp.com">>},{<<"id">>,<<"ASMS26FEOAG1">>},{<<"xmlns">>,<<"urn:xmpp:sid:0">>}],[]}]}},{offline_msg,{<<"918128184101">>,<<"im.myapp.com">>},{1533,664399,577631},never,{jid,<<"6-6001_115">>,<<"im.myapp.com">>,<<"A-6001-357871051876640">>,<<"6-6001_115">>,<<"im.myapp.com">>,<<"A-6001-357871051876640">>},{jid,<<"918128184101">>,<<"im.myapp.com">>,<<>>,<<"918128184101">>,<<"im.myapp.com">>,<<>>},{xmlel,<<"message">>,[{<<"from">>,<<"6-6001_115@im.myapp.com/A-6001-357871051876640">>},{<<"to">>,<<"918128184101@im.myapp.com">>},{<<"xml:lang">>,<<"en">>},{<<"id">>,<<"ExIDg-898">>},{<<"chat_sent">>,<<"2018-08-07T17:45:52.883Z">>},{<<"type">>,<<"chat">>}],[{xmlel,<<"ca">>,[],[]},{xmlel,<<"app_id">>,[],[{xmlcdata,<<"6001">>}]},{xmlel,<<"yek">>,[],[{xmlcdata,<<"vpib6chYuk29LSkZpw/B0GWV2HYfVtDeOsIha3uVOhq4p33GgPTb6nFgNPNcTdIjo1czGyBYqWrPLuY1kc9ZG6vx4DaEEjg/FNfoz31u9NBNFM+qb5BhSp6CDJ20yeb0gOdTWn6HkSeqThvmDPG3sdGH0mkDMgTHt29Pd+Uiqlg=">>}]},{xmlel,<<"yek_ym">>,[],[{xmlcdata,<<"M3iPJnBRDBT7rSTbFPBIovFZ0nmR355aHqV2q9fuoT4LT6hBmDRK2zDj60P9JEjYwCT2nRQCyKHqAu+5N4KIDjGgzc5lYkFOZZ8GiqhGEBtCDIqrUPJSEqw0b6KOJfaMrB5mv7tEHZJxBsVjG/UosVPwgXKIg6GdCLJ7SfP4zj4=">>}]},{xmlel,<<"subject">>,[],[{xmlcdata,<<"poke">>}]},{xmlel,<<"body">>,[],[{xmlcdata,<<"yzQJ7Q1P00ldi2wL1/0+cA==">>}]},{xmlel,<<"request">>,[{<<"xmlns">>,<<"urn:xmpp:receipts">>}],[]},{xmlel,<<"stanza-id">>,[{<<"by">>,<<"918128184101@im.myapp.com">>},{<<"id">>,<<"ASMS26UT15G1">>},{<<"xmlns">>,<<"urn:xmpp:sid:0">>}],[]}]}},{offline_msg,{<<"918128184101">>,<<"im.myapp.com">>},{1533,664401,594233},never,{jid,<<"6-6001_115">>,<<"im.myapp.com">>,<<"A-6001-357871051876640">>,<<"6-6001_115">>,<<"im.myapp.com">>,<<"A-6001-357871051876640">>},{jid,<<"918128184101">>,<<"im.myapp.com">>,<<>>,<<"918128184101">>,<<"im.myapp.com">>,<<>>},{xmlel,<<"message">>,[{<<"from">>,<<"6-6001_115@im.myapp.com/A-6001-357871051876640">>},{<<"to">>,<<"918128184101@im.myapp.com">>},{<<"xml:lang">>,<<"en">>},{<<"id">>,<<"ExIDg-901">>},{<<"chat_sent">>,<<"2018-08-07T17:45:53.439Z">>},{<<"type">>,<<"chat">>}],[{xmlel,<<"ca">>,[],[]},{xmlel,<<"app_id">>,[],[{xmlcdata,<<"6001">>}]},{xmlel,<<"yek">>,[],[{xmlcdata,<<"tVhR2yNtN3A/z2v5EP7uiE8DCAeSAzvx9VKLTAo7do/D6MaMUXOaQ9+qvT4QLLeH96MBrkKHr7kVfgWbfUH3+rRppOlRKyNoqPrp5FMjiWwjb/l1C0VQt6mhluMzLOiT5iRYtkXDKM4/CnQn9S6Jf2OJTMRugV1aivoXYtrHeX0=">>}]},{xmlel,<<"yek_ym">>,[],[{xmlcdata,<<"kf21oqFq0T00gUsJx8UyltFkDq9I/aHojfUdVEOIyatsgatsthGejRlrqc9Qf5vOO8Jz5cv9xRPLBd6Zgvt4HULXrDSrUgJWt+BveUQsOzD6V0M/6tZ5aR+XGnr84FoHCrnDANbazMxrfNRYvri8C/O+OZincqEMY0Q/AbHHtts=">>}]},{xmlel,<<"subject">>,[],[{xmlcdata,<<"poke">>}]},{xmlel,<<"body">>,[],[{xmlcdata,<<"WxqOqDAs4rvzIi9WAvAMNw==">>}]},...]}},...]}}`


Need an answer for both asap.

Thanks

**Comment by GalaxyGorilla** [comment]
The first issue looks pretty normal and I also have tons of them. That's just what you get when working in the mobile world due to network $things going on.

The second issue with offline message looks like a bug to me. Do those offline messages work at all?

**Comment by sandy5487** [comment]
@GalaxyGorilla 

this issue is nonpersistent in nature it comes up sometimes, and sometimes it does not. We unable to reproduce it purposefully and the root cause is still unknown.

**Comment by GalaxyGorilla** [comment]
What really baffles me is that this tuple from above (which contains the offline message) has a vastly different format than anything else handled in the c2s module (where the log is emitted). That's why I was curious if this is working at all.

**Comment by sandy5487** [comment]
@GalaxyGorilla 

Yes, offline messages are working fine but when this issue occurs the receiver got disconnected after getting the messages from the offline queue.

**Comment by sandy5487** [comment]
Hi @GalaxyGorilla @fenek @arcusfelis 

I have found one more warning log in mongooseIM logs:

17:11:25.264 [warning] no previous session with stream id <<"UDye3vcgEAHZjaAbMKZ5RViZzcHG">>

is that causing the disconnection of user?

Thanks

**Comment by GalaxyGorilla** [comment]
That's just a client attempting to resume an old session with stream management. Are those logs actually occuring together? If yes ... in which order? Do they belong to the same JID?

In situations like this I personally first do TCP dumps to check if the client in question behaves reasonable (what does he actually want?) and then I might trace the bug down in the application. Buy yea, that's kinda hard if you're not even able to reproduce this.

**Comment by sandy5487** [comment]
Hi all,

I have observed that when a client is trying to resume the current stream session this error occurs but the same scenario is handling perfectly in my previous Ejabberd server.

**Comment by fenek** [comment]
@sandy5487 

Maybe simply the client is able to reconnect after session times out? When you get `no previous session with stream id` warning, how much time passes between disconnection and resumption request?

**Comment by sandy5487** [comment]
it takes around 1-2 seconds

**Comment by sandy5487** [comment]
any update on this? @fenek

**Comment by fenek** [comment]
Can you please check if problems described in this issue are fixed by https://github.com/esl/MongooseIM/tree/c2s-resumption-edge-cases as well? I think they should be but there is a chance that these logs are triggered by yet another bug.

Thank you for your patience. :)

**Comment by sandy5487** [comment]
Hi @fenek 

we have tried again and while reproducing the same error we are facing the similar issue:
here is the error log.

when the user is getting offline messages this error occurs.

`
08:36:59.006 [info] Timeout on ejabberd_tls:send
08:36:59.006 [debug] Send element error: error, try enter resume session
08:36:59.006 [debug] targets [] from resources [{0,<<"ANDROID">>}] and ccenabled []
08:36:59.007 [debug] Send XML on stream = <<"<message from='sandeep@mg.myappdomain.com/Sandeeps-MacBook-Pro' to='918128184101@mg.myappdomain.com' type='chat' id='purple81f97925'><active xmlns='http://jabber.org/protocol/chatstates'/><body>3</body><stanza-id by='918128184101@mg.myappdomain.com' id='AT52UVU16901' xmlns='urn:xmpp:sid:0'/><delay xmlns='urn:xmpp:delay' stamp='2018-08-30T08:36:17.570833Z' from='mg.myappdomain.com'>Offline Storage</delay></message>">>
08:36:59.007 [debug] Error in ejabberd_tls:send: {error,enotconn}
08:36:59.007 [debug] Send element error: error, try enter resume session
08:36:59.007 [debug] targets [] from resources [{0,<<"ANDROID">>}] and ccenabled []
08:36:59.007 [debug] Send XML on stream = <<"<message from='sandeep@mg.myappdomain.com/Sandeeps-MacBook-Pro' to='918128184101@mg.myappdomain.com' type='chat' id='purple81f97926'><active xmlns='http://jabber.org/protocol/chatstates'/><body>4</body><stanza-id by='918128184101@mg.myappdomain.com' id='AT52V01LIU81' xmlns='urn:xmpp:sid:0'/><delay xmlns='urn:xmpp:delay' stamp='2018-08-30T08:36:18.047349Z' from='mg.myappdomain.com'>Offline Storage</delay></message>">>
08:36:59.007 [debug] Error in ejabberd_tls:send: {error,enotconn}
08:36:59.007 [debug] Send element error: error, try enter resume session
08:36:59.007 [debug] targets [] from resources [{0,<<"ANDROID">>}] and ccenabled []
08:36:59.007 [debug] Send XML on stream = <<"<message from='sandeep@mg.myappdomain.com/Sandeeps-MacBook-Pro' to='918128184101@mg.myappdomain.com' type='chat' id='purple81f97927'><active xmlns='http://jabber.org/protocol/chatstates'/><body>5</body><stanza-id by='918128184101@mg.myappdomain.com' id='AT52V04N62G1' xmlns='urn:xmpp:sid:0'/><delay xmlns='urn:xmpp:delay' stamp='2018-08-30T08:36:18.447283Z' from='mg.myappdomain.com'>Offline Storage</delay></message>">>
`

**Comment by GalaxyGorilla** [comment]
So, here is a half educated wild guess:

The 120 milliseconds from the [TCP listener config](https://github.com/esl/MongooseIM/blob/master/src/mongoose_tcp_listener.erl#L114) are just not enough to cope with your network conditions. The socket closes the connection after timeout, because `send_timeout_close` is `true`, which explains the `enotconn` messages.

120 milliseconds is not much and unfortunately this is hardcoded. Can you change that number yourself to sth of the order of 5000 and test it out?

**Comment by fenek** [comment]
Oh, there is a good chance that it's a simple mistake and someone intended to set it to 120 seconds. :) Worth fixing.

**Comment by fenek** [comment]
It has been changed in https://github.com/esl/MongooseIM/commit/01e494a90d06c89b15b2c09a927d899a7b22dc84#diff-c6c5aab86dcf2b2dc7c51efcbdc54ba1 - a very good chance of a simple mistake. Unfortunately it is not overridable. I'll create an internal ticket for this.

**Comment by GalaxyGorilla** [comment]
I already wondered why the ?TCP_SEND_TIMEOUT macro is defined in the listener module and then used somewhere else in the s2s module ;))

**Comment by sandy5487** [comment]
Thanks, @GalaxyGorilla, @fenek   I have updated the source code and its working fine now :)

**Comment by fenek** [comment]
Fixed by #2069
